### PR TITLE
[FEAT][DMA] add DMA graph for DMAI/DMAO

### PR DESCRIPTION
**Pull request by thoangtrvn**
This PR is to address the issue #53 

In this PR: 
* new APIs to generate DCI for 1D, 2D and 3D tensors with generic stick layout on the device
* a new DMA graph to perform data copy in and out of Spyre device
* a new API to detect memory needs on Spyre 

In this PR, the DCIs generated support:
- Generic stick format for 1, 2, and 3D tensors
- Padding the size of the stick dimensions up to a size of 64

When we invoke 
```python
x = torch.rand(128, 64, dtype=torch.float16)
y = x.to('spyre')
```
DCI (Data Conversion Information) provides a mechanism to map elements from a CPU tensor to elements in a data layout on Spyre. Here, no implicit dimension swapping is performed with this DCI code.

NOTE: 4D DCI will be added soon.

### CODE REVIEW

File: tests/test_modules.py [hunk 1]
```diff
@@ -40,6 +40,7 @@ def __init__(self, method_name="runTest", methodName="runTest"):
         self.atol = 1e-3
         self.dtype = torch.float16
 
+    @unittest.skip("Swapping stick dimension is unsupported in new DCI")
```
Comment (ani300) [inline tests/test_modules.py]: mark this as a TODO

Comment (ani300) [reply tests/test_modules.py]: or at least open issue to support it

File: torch_spyre/csrc/spyre_mem.cpp [hunk 2]
```diff
@@ -44,6 +48,411 @@
...
 namespace spyre {
 
+struct DMAParameters {
+  const int64_t size_bytes;  // bytes to transfer
+  const off64_t src_offset;
+  const off64_t dst_offset;
+};
+auto get_device_shape(c10::IntArrayRef sizes) -> std::vector<int64_t> {
+  /* Given the CPU Tensor's shape, return the shape of the tensor on Spyre
+   */
```
Comment (ani300) [inline torch_spyre/csrc/spyre_mem.cpp]: this assumes the tensor on cpu is contiguous right? this is why we ignore the strides completely. If so, please make it clear on the comment

File: torch_spyre/csrc/spyre_mem.cpp [hunk 3]
```diff
@@ -25,8 +25,12 @@
 #include <c10/core/TensorOptions.h>
 #include <c10/util/ArrayRef.h>
 #include <torch/library.h>
+#include <util/sen_data_convert.h>
```
[the remark this answers is not in the document]
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: that's correct. Maybe it should be migrated to runtime or this repo in the future?

[the remark this answers is not in the document]
Comment (avery-blanchard) [reply torch_spyre/csrc/spyre_mem.cpp]: Yes, this is a deeptools header. All of the data conversion operations are in `deeptools/utils` and called by Flex to handle the SenHostCompute nodes. Eventually, I think `utils` should be moved.

File: torch_spyre/csrc/spyre_mem.cpp [hunk 4]
```diff
@@ -44,6 +48,411 @@
...
+      dev_shape.push_back(stick_size);
+      if (requires_padding) {
+        dev_shape.push_back(1);
+      } else {
+        dev_shape.push_back(cpu_shape.back() / stick_size);
+      }
+    } else if (i == 0) {
+      dev_shape.push_back(stick_size);
+    } else if (i == sizes.size() - 1) {
+      if (cpu_shape.size() <= 2) dev_shape.push_back(cpu_shape[i - 1]);
```
Comment (ani300) [inline torch_spyre/csrc/spyre_mem.cpp]: put brackets on the if even if it's just 1 statement

File: torch_spyre/csrc/spyre_mem.cpp (hunk 2 above)
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: Yes. @ani300 : maybe we should give a test case for non-contiguous tensor, and marked as skipped so that we can test that in the future. Do you have one suggestion of such tensor?

File: torch_spyre/csrc/spyre_mem.cpp [hunk 5]
```diff
@@ -44,6 +48,411 @@
...
+      if (requires_padding) {
+        auto padded_shape =
+            cpu_shape.back() + (stick_size - (cpu_shape.back() % stick_size));
+        dev_shape.push_back(padded_shape / stick_size);
+      } else {
+        dev_shape.push_back(cpu_shape.back() / stick_size);
+      }
+      if (cpu_shape.size() > 2) dev_shape.push_back(cpu_shape[i - 1]);
```
Comment (ani300) [inline torch_spyre/csrc/spyre_mem.cpp]: if I understand this correctly, if the cpu tensor only has 2 dimensions, then you're doing [stick_size, cpu_dim[0], cpu_dim[1]/stick_size], while on dim > 2 (e.g. 3): [stick_size, cpu_dim[0], cpu_dim[2]/stick_size, cpu_dim[1]]?

If this is the case, please add a comment explaining the logic and also add tests somewhere for all the possible cases you're running through here.

File: torch_spyre/csrc/spyre_mem.cpp [hunk 6]
```diff
@@ -44,6 +48,411 @@
...
+  return dev_shape;
+}
+auto get_device_shape(const at::Tensor* tensor) -> std::vector<int64_t> {
+  /* Given the CPU Tensor, return the shape of the equivalent tensor on
+   * Spyre
+   */
+  const c10::IntArrayRef& sizes = tensor->sizes();
+  return get_device_shape(sizes);
+}
+auto get_device_layout(c10::IntArrayRef sizes, c10::IntArrayRef strides)
```
Comment (ani300) [inline torch_spyre/csrc/spyre_mem.cpp]: strides are ignored here, right? put it in a comment and maybe make a TODO to support strided tensors later

File: torch_spyre/csrc/spyre_mem.cpp [hunk 7]
```diff
@@ -44,6 +48,411 @@
...
+      auto fdc =
+          fgb.SenFusedDeviceCompute("FDC", {dci_ti}, {inp_node}, fdc_graph);
+      auto d2h_dci = generate_dci(dev_tensor, host2device);
+      sendnn::NodePtr d2h_dci_node = fgb.SenHostCompute(
+          "Sen2Host-HostPrep", cpu_ti, fdc, "SenDataConvert", d2h_dci);
+
+      fgb.PrimaryOutput("SN PO", d2h_dci_node->OutputPort(0));
+    }
+
+    SEN_THROW_NOK(fgb.Finalize(&exec_graph));
```
Comment (ani300) [inline torch_spyre/csrc/spyre_mem.cpp]: variable names need a pass in this function

**Review by ani300 — CHANGES_REQUESTED** [review]
This PR needs a lot more documentation explaining some of the choices. Plus, I think some of the logic is suspicious when the stickfied dimension(s) is > stick_size

File: torch_spyre/csrc/spyre_mem.cpp [hunk 8]
```diff
@@ -44,6 +48,411 @@
...
+                            sendnn::TensorLocation::DEVICE());
+  sendnn::TensorInfo dci_ti(sendnnSpyreT, dev_tensor_shape, layout,
+                            sendnn::TensorLocation::HOST());
+  //  STAGE 1: execution graph
+  sendnn::SubGraph fdc_graph;
+  int64_t xfer_size = dev_tensor_shape.Volume() * cpu_tensor->element_size();
+  {  // subgraph (execution graph)
+    flex::FlexGraphBuilder gb;
+    flex::FlexGraphBuilder* gb_sn = &gb;
+    DMAParameters tp{xfer_size, 0, 0};  // (num_bytes, offset_src, offset_dst)
```
[the remark this answers is not in the document]
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: tp = transfer-parameter. Do you have a suggestion for a better name, @ani300 ?

File: torch_spyre/csrc/spyre_mem.cpp [hunk 9]
```diff
@@ -44,6 +48,411 @@
...
+  sendnn::TensorInfo dev_ti(sendnnSpyreT, dev_tensor_shape, layout,
+                            sendnn::TensorLocation::DEVICE());
+  sendnn::TensorInfo dci_ti(sendnnSpyreT, dev_tensor_shape, layout,
+                            sendnn::TensorLocation::HOST());
+  //  STAGE 1: execution graph
+  sendnn::SubGraph fdc_graph;
+  int64_t xfer_size = dev_tensor_shape.Volume() * cpu_tensor->element_size();
+  {  // subgraph (execution graph)
+    flex::FlexGraphBuilder gb;
+    flex::FlexGraphBuilder* gb_sn = &gb;
```
[the remark this answers is not in the document]
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: flex::FlexGraphBuilder is the derived class of sendnn::GraphBuilder and provides additional APIs that sendnn::GraphBuilder doesn't have.

File: torch_spyre/csrc/spyre_mem.cpp (hunk 7 above)
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: @ani300 : we revised the names.

File: torch_spyre/csrc/spyre_mem.cpp [hunk 10]
```diff
@@ -44,6 +48,411 @@
...
+
+      fgb.PrimaryOutput("SN PO", d2h_dci_node->OutputPort(0));
+    }
+
+    SEN_THROW_NOK(fgb.Finalize(&exec_graph));
+  }
+
+  sendnn::SegmentTable segment_table = {
+      sendnn::Segment::PRIMARY_OUT(xfer_size),
+      sendnn::Segment::PRIMARY_IN(xfer_size),
```
[the remark this answers is not in the document]
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: this is the constraint as we're using SenFusedDeviceNode pathway.

File: torch_spyre/csrc/spyre_mem.cpp [hunk 11]
```diff
@@ -44,6 +48,411 @@
...
+
+      auto fdc =
+          fgb.SenFusedDeviceCompute("FDC", {dci_ti}, {h2d_dci_node}, fdc_graph);
+      fgb.PrimaryOutput("SN PO", fdc->OutputPort(0));
+    } else {
+      sendnn::NodePtr inp_node = fgb.PrimaryInput(
+          "SN PI", dci_ti);  // not important in terms of shape (as it will be
+                             // empty when using)
+      // sendnn::NodePtr inp_node = fgb.HeapInput("SN PI", dci_ti); //not
+      // important in terms of shape (as it will be empty when using)
```
[the remark this answers is not in the document]
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: removed

File: tests/test_ops.py [hunk 12]
```diff
@@ -53,29 +53,60 @@ def test_inplace_fill_scalar(self):
...
+        y = x.to("spyre").to("cpu")
+        torch.testing.assert_close(y, x, rtol=self.rtol, atol=self.atol)
+
     def test_t_1d(self):
         x = torch.tensor([1, -2, 3], dtype=self.dtype)
         x_spyre = x.to("spyre")
         y = x_spyre.t().to("cpu")
         torch.testing.assert_close(y, x.t(), rtol=self.rtol, atol=self.atol)
 
+    @unittest.skip("Swapping stick dimension is unsupported in new DCI")
```
Comment (JRosenkranz) [inline tests/test_ops.py]: Do we plan to eventually support this? We may want to mention it will be supported in a later release

Comment (thoangtrvn) [reply tests/test_ops.py]: @JRosenkranz : this is mentioned in another comment by @avery-blanchard . she will add a dedicated issue for this; and it is not part of current DMA scope.

File: torch_spyre/csrc/spyre_mem.cpp (hunk 7 above)
Comment (ani300) [reply torch_spyre/csrc/spyre_mem.cpp]: there's still a lot of 1 character/2 character/3 character variable names. Please make variable names longer and more descriptive

File: tests/test_modules.py (hunk 1 above)
Comment (ani300) [reply tests/test_modules.py]: is the issue open on tihs? please link here when you mark as resolved

File: torch_spyre/csrc/spyre_mem.cpp [hunk 13]
```diff
@@ -44,6 +48,411 @@
...
+  auto stick_size = 64;
+  auto requires_padding = (cpu_shape.back() % stick_size != 0);
+
+  for (int i = 0; i < cpu_shape.size(); i++) {
+    if (i == 0 && cpu_shape.size() == 1) {
+      dev_shape.push_back(stick_size);
+      if (requires_padding) {
+        dev_shape.push_back(1);
+      } else {
+        dev_shape.push_back(cpu_shape.back() / stick_size);
```
[the remark this answers is not in the document]
Comment (avery-blanchard) [reply torch_spyre/csrc/spyre_mem.cpp]: Tensors where the size of the stick dimension is not a multiple of the stick size will require padding. They should not hit this. I can add comments about this.

File: torch_spyre/csrc/spyre_mem.cpp (hunk 2 above)
Comment (thoangtrvn) [reply torch_spyre/csrc/spyre_mem.cpp]: @ani300 : we added a TODO to support strides tensors.
